assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * DashboardAllTrafficWidget component.
+ *
+ * Site Kit by Google, Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { STORE_NAME } from '../../datastore/constants';
+import { STORE_NAME as CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
+import { STORE_NAME as CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
+import { trafficSourcesReportDataDefaults } from '../../util';
+import whenActive from '../../../../util/when-active';
+import ErrorText from '../../../../components/error-text';
+import AcquisitionPieChart from '../common/AcquisitionPieChart';
+import AcquisitionSources from '../common/AcquisitionSources';
+const { useSelect } = Data;
+
+function DashboardAllTrafficWidget() {
+	const { report, error } = useSelect( ( select ) => {
+		const store = select( STORE_NAME );
+		const args = {
+			...trafficSourcesReportDataDefaults,
+			dateRange: select( CORE_USER ).getDateRange(),
+		};
+
+		const url = select( CORE_SITE ).getCurrentEntityURL();
+		if ( url ) {
+			args.url = url;
+		}
+
+		return {
+			report: store.getReport( args ),
+			error: store.getErrorForSelector( 'getReport', [ args ] ),
+		};
+	} );
+
+	if ( error ) {
+		return (
+			<div className="mdc-layout-grid__cell mdc-layout-grid__cell--span-12">
+				<ErrorText message={ error.message } />
+			</div>
+		);
+	}
+
+	return (
+		<div className="mdc-layout-grid__cell mdc-layout-grid__cell--span-12">
+			<div className="mdc-layout-grid__inner">
+				<div className="mdc-layout-grid__cell mdc-layout-grid__cell--span-4-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-4-phone">
+					<AcquisitionPieChart data={ report } source />
+				</div>
+				<div className="mdc-layout-grid__cell mdc-layout-grid__cell--span-8-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-4-phone">
+					<AcquisitionSources data={ report } />
+				</div>
+			</div>
+		</div>
+	);
+}
+
+export default whenActive( { moduleName: 'analytics' } )( DashboardAllTrafficWidget );
```

Comment:
Why use the HOC component here? I'd prefer we not introduce any new HOCs into the codebase, as we're using hooks going forward and HOCs are (in my view) more cognitive load to understand and work on.

Why not use the selector logic in the `whenActive` HOC here? https://github.com/google/site-kit-wp/pull/1931/files#diff-451a45872b732c4c542fba8c369ba0b2R44

I don't think it's worth abstracting that behaviour into its own HOC, as I find it harder-to-follow the abstracted approach (it's less direct and requires extra code) compared to a single `useSelect` in the component that, if `false` allows the component to return a fallback/`null`.

Reply:
We need the HOC because it helps us to encapsulate the logic for detecting module status and let us get rid of duplicating the logic across a dozen of different widgets. In other words it lets us control when to display a widget without breaking DRY principle.

Reply:
The DRY you mention is pretty straightforward, and much easier to understand without the HOC obscuring the logic of the component. The main code you're abstracting is:

```js
const moduleInfo = useSelect( ( select ) => select( MODULES_STORE ).getModule( moduleName ) );

if ( ! moduleInfo ) {
    return null;
}

if ( ! moduleInfo.active || ! moduleInfo.connected ) {
    return <div>Fallback</div>;
}
```

Repeating that sort of code inside components/modules is very straightforward and I don't think worth adding the HOC. An issue with the HOC approach is that it moves some of the logic handling of a component _outside_ the body of the component, so understanding the component's flow and potential return values becomes more complex and spread-out. In particular, here, one needs to read:

* https://github.com/google/site-kit-wp/blob/166405fdfb075c0447d06352dbc37b0fafb3ed3c/assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget.js#L33-L72
* https://github.com/google/site-kit-wp/blob/166405fdfb075c0447d06352dbc37b0fafb3ed3c/assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget.js#L74
* https://github.com/google/site-kit-wp/blob/166405fdfb075c0447d06352dbc37b0fafb3ed3c/assets/js/util/when-active.js#L42-L65

to understand a reasonably simple/concise conditional.

Reducing repetition of the
```js
if ( ! moduleInfo ) {
    return null;
}

if ( ! moduleInfo.active || ! moduleInfo.connected ) {
    return <div>Fallback</div>;
}
```
code could be done with a hook that returned an object with `{ moduleInfo, activeAndConnected }`.

If you want to consolidate this code, I think a custom hook would be better.

But right now this is only being used in one place, and the fallback component you've implemented isn't being used. I don't think we should be preemptively optimising/DRYing up code that isn't yet being repeated. I think it would be better to see if, in future modules, we need to compare active-vs-connected or if checking for both always returns the same fallback. If we want non-null for lacking modules, etc. Let's abstract _after_ we've built a few of these components, as we don't know how they'll go yet. Given there wasn't anything in the IB about creating this abstracting and the selectors involved are only a few lines. In the case of this component:

```js
if ( ! moduleInfo ||  ! moduleInfo.active || ! moduleInfo.connected  ) {
    return null;
}
```

will be fine.

So I think we should remove the HOC approach and leave it in the component itself for now. If we do want to abstract in the future, let's prefer custom hooks over HOCs.

Reply:
It's a very good point, @aaemnnosttv! Thanks for raising it.

Reply:
I had considered the "return early" benefit of the HOC, but I think that should be thought of as a convenience, not a requirement. In the past we had selectors that would error or require particular arguments, but found that created problems like this—in the end the approach we've been taking is that selectors are always safe to call, and will just return `undefined` if not supplied with the right arguments. I don't _think_ we have any selectors left that are unsafe, but if we do: we should fix that. Even if not for this issue, they're in general tough to work with.

In terms of the performance hit of running many selectors… selectors are pretty fast to run so I don't think that's a huge deal. It _could_ matter if resolvers that make HTTP requests are involved, but if they're getting invalid/empty arguments (a bunch of `undefined` because a component's data isn't ready) then they should be designed to skip useless HTTP calls.

The main point of exiting early is neat, but there is a substantial cognitive overhead to higher-order-components compared to hooks. For newer developers hooks are easier to understand and HOCs can be a real pain when they're layered/composed.

From my point-of-view, there's no technical reason not to use hooks, and doing so means our components have a lower surface area of knowledge needed and are more consistent. I also think it's easier to understand a component's logic when its entirely encapsulated within the component. That's a big reason for using hooks over HOCs in general, so I'd prefer we use them here.

If there are reasons _right now_ that mean we can't—certain selectors that cause errors when run without arguments, for instance—then the HOC approach is fine, at least as a stopgap. But I don't see any reference to hooks being technically impossible to use, just the preference for HOCs and the "return early" approach. I understand that might be a slightly more efficient code path, but I think it's also a harder-to-follow code path. I'd rather our code be ever-so-slightly less optimised but more maintainable, and I think eschewing HOCs whenever possible gives us that 🙂 

-----

Regarding `isModuleConnected` though: yes, let's add it. Sounds good 👍 . I think doing so elsewhere/in a new issue might be easier, but whatever works for others 🙂

Reply:
@tofumatt I feel you are fighting with a non-existing problem. There is no cognitive overhead in HOCs. It's a very simple pattern that is widely used and much easier to understand than even hooks.

Can you tell exactly what makes it hard for you to understand it? A need to look into two files to understand how a component works? We use similar approach in datastores when we use higher-order functions (like createFetchStore) to generate different pieces of a datastore and it is fine for everyone.

Anyway, IB is approved and it states that we need to use whenActive HOC, so let's stick with it. If you want, create a new issue to replace whenActive with a hook. Ok?

Reply:
`isModuleConnected` is added.

Reply:
After another look and some clarifications from @aaemnnosttv, the use of HOCs, particularly with selectors that have resolvers, is a good call here. Let's stick with using them, sounds good.

One of the main reasons to use them over hooks here is that hooks run with resolvers may call out to modules that aren't active, which would result in a lot of 404s at best and errors at worst. It might be worth adding another comment about that to clarify their usage, as we generally use render-props pattern or hooks over HOCs in the codebase. But I'm happy to stick with HOCs here 👍